src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,353 @@
+@file:Suppress("MemberVisibilityCanBePrivate")
+
+package app.revanced.patcher.apk
+
+import app.revanced.patcher.DomFileEditor
+import app.revanced.patcher.Patcher
+import app.revanced.patcher.PatcherOptions
+import app.revanced.patcher.arsc.*
+import app.revanced.patcher.util.ProxyBackedClassList
+import app.revanced.patcher.util.xml.LazyXMLInputSource
+import app.revanced.patcher.util.xml.scanIdRegistrations
+import com.reandroid.apk.ApkModule
+import com.reandroid.apk.xmlencoder.EncodeException
+import com.reandroid.apk.xmlencoder.EncodeMaterials
+import com.reandroid.apk.xmlencoder.EncodeUtil
+import com.reandroid.archive.InputSource
+import com.reandroid.arsc.chunk.PackageBlock
+import com.reandroid.arsc.chunk.TableBlock
+import com.reandroid.arsc.chunk.xml.AndroidManifestBlock
+import com.reandroid.arsc.value.Entry
+import com.reandroid.arsc.value.ResConfig
+import lanchon.multidexlib2.BasicDexEntry
+import lanchon.multidexlib2.DexIO
+import lanchon.multidexlib2.MultiDexContainerBackedDexFile
+import lanchon.multidexlib2.MultiDexIO
+import lanchon.multidexlib2.RawDexIO
+import org.jf.dexlib2.dexbacked.DexBackedDexFile
+import org.jf.dexlib2.iface.MultiDexContainer
+import org.jf.dexlib2.writer.io.MemoryDataStore
+import org.w3c.dom.*
+import java.io.File
+import java.util.zip.ZipEntry
+
+
+/**
+ * The [Apk] file that is to be patched.
+ *
+ * @param path The path to the apk file.
+ * @param name The name of this apk.
+ */
+sealed class Apk private constructor(internal val module: ApkModule) {
+    companion object {
+        const val manifest = "AndroidManifest.xml"
+    }
+
+    /**
+     * The metadata of the [Apk].
+     */
+    val packageMetadata = PackageMetadata(module.androidManifestBlock)
+
+    internal val archive = Archive(module)
+
+
+    /**
+     * Refresh updated resources for an [Apk].
+     *
+     * @param options The [PatcherOptions] of the [Patcher].
+     */
+    internal open fun finalize(options: PatcherOptions) {
```

Comment:
Should this not be close from Closeable?

Reply:
Doesn't `close` imply that you are not allowed to use it again?

You still have to call `Apk.write()` after `Apk.finalize/close`

Reply:
Ideally close is called right before write, or it writes on close. Why would you want to refresh the resources multiple times?

Reply:
What if we just finalize resources in `write()`?

That might make it harder to track progress for manager, but that is not a big problem and can be resolved later

Reply:
> What if we just refresh resources in `write()`?

Thats what I ended up going with I guess. Makes sense to me.